tests/test_cf/test_unit/test_template.py

```diff
@@ -285,3 +290,163 @@ def test_render_condition_keys():
+            "/account/555555555555/short_name": "cld-55-rdr",
+        }
+    }
+
+    template = Template(t, dynamic_references=dynamic_references)
+
+    result = template.render()
+
+    # This item "just" resolves the SSM parameter
+    foo_resource_props = result["Resources"]["Foo"]["Properties"]
+    assert foo_resource_props["PolicyName"] == "mgt-cld-rdr-launch-role-pol"
+
+    # This item resolves the SSM parameter after a substitution has been performed
+    bar_resource_props = result["Resources"]["Bar"]["Properties"]
+    assert bar_resource_props["PolicyName"] == "mgt-cld-55-rdr-launch-role-pol"
+
+    # This item contains multiple resolved parameters
+    two_resource_props = result["Resources"]["TwoRefTest"]["Properties"]
+    assert two_resource_props["PolicyName"] == "mgt-cld-55-rdr-launch-cld-rdr-pol"
+
+
+def test_resolve_all_types_dynamic_references():
+    # This contains one of each type of dynamic references, with
+    # the example resources coming from the documentation
+    # https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/dynamic-references.html
+    t = {
+        "Resources": {
+            "MyRDSInstance": {
+                "Type": "AWS::RDS::DBInstance",
+                "Properties": {
+                    "DBName": "MyRDSInstance",
+                    "AllocatedStorage": "20",
+                    "DBInstanceClass": "db.t2.micro",
+                    "Engine": "mysql",
+                    "MasterUsername": (
+                        "{{resolve:secretsmanager:MyRDSSecret:SecretString:username}}"
+                    ),
+                    "MasterUserPassword": (
+                        "{{resolve:secretsmanager:MyRDSSecret:SecretString:password}}"
+                    ),
+                },
+            },
+            "MyIAMUser": {
+                "Type": "AWS::IAM::User",
+                "Properties": {
+                    "UserName": "MyUserName",
+                    "LoginProfile": {
+                        "Password": "{{resolve:ssm-secure:IAMUserPassword:10}}"
+                    },
+                },
+            },
+            "MyS3Bucket": {
+                "Type": "AWS::S3::Bucket",
+                "Properties": {"AccessControl": "{{resolve:ssm:S3AccessControl:2}}"},
+            },
+        }
+    }
+
+    dynamic_references = {
+        "ssm": {"S3AccessControl:2": "private"},
+        "ssm-secure": {"IAMUserPassword:10": "my-really-secure-iam-password"},
+        "secretsmanager": {
+            "MyRDSSecret:SecretString:username": "my-username",
+            "MyRDSSecret:SecretString:password": "my-password",
+        },
+    }
+
+    template = Template(t, dynamic_references=dynamic_references)
+    result = template.render()
+
+    rds_resource_props = result["Resources"]["MyRDSInstance"]["Properties"]
+    assert rds_resource_props["MasterUsername"] == "my-username"
+    assert rds_resource_props["MasterUserPassword"] == "my-password"
```

Comment:
I wouldn't change this, this test is fine. But since you are a user of the library, I wanted to tell you about #210. Which adds some helper functions.

You can now do `result = template.create_stack()`. Which returns a "stack" object. It's also backward compatible. So even if you only changed that line, the rest of your tests would still work because the "stack" object is a dictionary as well. 

The stack object can also help you get resources, conditions, outputs etc.

`rds_resource = stack.get_resource("MyRDSInstance")`

Under the hood the stack object is doing `asserts` and making sure these things exist. It returns "wrapper" objects that have additional helper/testing methods.

`rds_resource.assert_property_has_value("MasterUsername", "my-username")`

it might be time I create a static website for documentation 🤔

Reply:
I had seen that actually in the sample test case & used it when I was first playing with the library. I think I've got some suggestions of helpful functions to add to that too, but 2 discussions/PRs is enough for now :)

For this, I'd just copied the pattern from the other tests without realising the two approaches were equivalent.